### DISCUSSION TITLE
forwarding to new url adds "/"

### DISCUSSION THREAD

**Discussion by UmbrellaTester** [cypress-io/cypress#24817] [Questions and Help]
Hi everyone,

in the application i am testing is the Login Process a fowarding to our internal ID Server.
If i do the test manually its fine but in Cypress the new url gets an additional "/".

My Code is:

context("Login", () => {
  
    it('should ensure user is logged out from id server', () => {
      cy.visit(idserver), {timeout: 20000};
      cy.get('body').wait(5000)
        .then(($body) => {
          if ($body.find('[data-cy=burgerMenu]', {timeout: 8000}).length) {
            cy.get('[data-cy=burgerMenu]').click();
            cy.get('[data-cy=option_logout]', {timeout: 8000}).click();
          }
  
        });
      cy.wait(5000);
      
    });
  
    it ('should login via ID-Server', () => {
      cy.visit('/login', {timeout: 30000});
      cy.url().should('contain', '/login');
      cy.get('[data-cy=email]').should('be.visible');
      cy.get('[data-cy=email]').type(email);
      cy.get('[data-cy=password]').should('be.visible');
      cy.get('[data-cy=password]').type(password);
      cy.get('[data-cy=login]').click();
    })
  })

After calling my application the application is forwarding to our internal ID Server. The URL then includes the string #forwardedFromCloud.

in the Cypress Logs you can see that the right url is called but after the forwarding to the application the url contains an additional "/". The url now calls a website that is not complete and the test fails.

![image](https://user-images.githubusercontent.com/119041936/203949133-d3a6b848-cd9e-4534-be76-e67310f5dc10.png)

I talked many times to our devs if there is any possibilty that the new url gets the additional "/" from their code but there is no possible way that this could happen. 

If i call the the new url (with the additional "/") in a web browser outside of cypress its perfectly fine because the browser removes it.
In Cypress this doesnt happen. Is there a way to make the browser called by cypress to remove it too?

Or is there any other solution to that?

**Comment by portyu9** [comment]
Hi @UmbrellaTester 

Cypress is most likely not adding that `/`.

The change:

```text
#forwardedFromCloud
```

to:

```text
#/forwardedFromCloud
```

is typical of a client-side router using hash routing. For example, Angular's `HashLocationStrategy` normalizes routes to `#/...`.

So the fix should be in the application/redirect URL, not Cypress.

If `forwardedFromCloud` is only a marker and not an application route, move it out of the hash and use a query parameter instead:

```text
/login?...&forwardedFromCloud=true
```

If it is intended to be a hash route, then make the application accept:

```text
#/forwardedFromCloud
```

and configure that route accordingly.

Also, because the login flow moves to another origin, current Cypress tests should use `cy.origin()` for commands executed on the ID-server origin.

There is no Cypress setting that should be used to "remove" the `/`; the URL/router logic should generate and handle the correct callback URL.

Please mark this answer as accepted if it helped.
